### PR TITLE
12188 add logging handler context managers

### LINKED ISSUE #12188: Add a context manager for calling `log.failure` when code fails
Right now we have a pattern of doing this:

```python3
log = Logger()

try:
    applicationCode(parameter)
except BaseException:
    log.failure("while doing xyz with {parameter}", parameter=parameter)
else:
    proceed()
```

This is verbose, error-prone, and kinda reads backwards.  Instead, something like this would be nicer:

```python3
log = Logger()
with log.whenThereIsAnyProblemLogThis("while doing xyz with {parameter}", parameter=parameter) as operation:
    applicationCode(parameter)
if operation.succeeded:
    proceed()
```

### LINKED ISSUE #12195: demonstrate the utility of / migrate to `Logger.handlingFailures`
#12188 describes a new API for handling failures, as is often done within Twisted itself. So let's demonstrate doing this within Twisted itself!

### PR DESCRIPTION
**Pull request by glyph**
This is a retread of #12196 with a simpler commit history. Its description:

> ## Scope and purpose
> 
> Fixes #12188
> 
> Fixes #12195 
> 
> This includes all the changes from #12189 , but attempts to demonstrate their utility. In so doing, I discovered a performance regression due to the additional allocation from creating a new context manager, so this also adds `Logger.failureHandler`, which allows for pre-allocating a context manager with a *fixed* error message, suitable for use within hot loops.
> 
> It also includes a ton of examples of the usages of both decorators, as well as a *huge* amount of refactoring, because:
> 
> - many of the places that I wanted to use this are *very* old Twisted code
>   - `ThreadedSelectReactor` was, in particular, a buggy mess with no test coverage that basically didn't work
> - using this new API requires migrating from old-logger to new-logger, so I had to go do that too
> - as I migrated I wanted type checking, and touching type checking in some places caused knock-on changes to be needed
> - in addressing the coverage gaps, I fixed a few bugs, in particular in half-close support, which previously just … did not work at all, with `StandardIO` just never calling `writeConnectionLost` at all.
> 
> There are no particularly interesting bugs fixed here, or major new subsystems with full type coverage, so not much for the release notes, but i think it's worth dealing with a slightly larger and more diffuse PR for getting all these minor improvements.

### CODE REVIEW

**Comment by glyph** [comment]
@twm fyi this is your fault for making me actually consider the design

(Thank you, the code is much better for it, but still, there's like 1000 lines now)

**Comment by glyph** [comment]
Sigh, trying one more force push to get rid of spurious codecov statuses that seem to have gotten stuck to the win32 stdio changes.

**Comment by adiroiban** [comment]
> Sigh, trying one more force push to get rid of spurious codecov statuses that seem to have gotten stuck to the win32 stdio changes.

Not sure if you are aware of the "Re-run failed jobs" feature.

Maybe a force push is easier for your, so that's also an option.

image

File: docs/core/howto/logger.rst
Comment (adiroiban) [inline docs/core/howto/logger.rst]: Just a suggestion.
To make the name closer to the intended usage

> Note that this API is meant to capture **unexpected and otherwise unhandled** errors (in other words: bugs, which is why tracebacks are preserved).

```suggestion
    with log.handleUnexpectedError("While doing some math:"):
```

File: src/twisted/internet/base.py
Comment (adiroiban) [inline src/twisted/internet/base.py:82]: Maybe document why we have these:

```suggestion
# Pre-allocating contexts manager with fixed error messages, suitable for use within hot loops.
_topHandler = _log.failureHandler("Unexpected error in main loop")
```

File: docs/core/howto/logger.rst
Comment (adiroiban) [inline docs/core/howto/logger.rst]: Maybe we should also have an example demonstrating the usage of the "Operation" ... or at least a few words about the fact that the context manager returns an "operation" that can be used to check the result and contains a reference to the raised Failure

**Review by adiroiban — APPROVED** [review]
Thanks for the update.

I did a quick review and left a few comments.

I haven't looked into the details of `win32` and `threadselect` changes 

I think that this is a big improvement.

I have approved it, but I am leaving this for review for Tom or other team members.

If a further review is delayed, I think that this can be merged.

Thanks again!

**Comment by glyph** [comment]
Thanks @adiroiban, I really appreciate that! I think this is one that might benefit from some post-hoc review, so after reviewing and addressing your comments I might land pretty quickly, but will be interested in feedback from others later.

**Comment by glyph** [comment]
> Not sure if you are aware of the "Re-run failed jobs" feature.
> 
> Maybe a force push is easier for your, so that's also an option.

The problem in this case is that the job did not fail, but that codecov reported annotations 

> Not sure if you are aware of the "Re-run failed jobs" feature.
> 
> Maybe a force push is easier for your, so that's also an option.

I use it all the time :)

In this particular case, the issue is that the job did not actually fail. Codecov ran, it emitted some annotations, failed the check, then withdrew the failure once the Windows coverage arrived, but *didn't* withdraw the annotations.  You can see them on the current files changed tab like this: 

image

which of course makes no sense given that the patch is 100% covered and these are patch lines.  I was hoping that the bug was intermittent and might be addressed by a full rebuild, but no, it looks like a flaw in codecov.

**Comment by glyph** [comment]
> I have approved it

huh, why did the robot not notice this and add `needs-merge` ?

File: docs/core/howto/logger.rst
Comment (glyph) [reply docs/core/howto/logger.rst]: I thought for a really long time about this, and I think I am going to reject the suggestion on the basis of length; this is already pretty wordy, and we put all the words in to explain its purpose (`with log.handlingUnexpectedErrorsWithMessageAndContinuing`) it gets unwieldy, and "unexpected" is a slippery word to put in a place as definitive as a name. (After all, we *kinda* expect it here, it's the application code which doesn't expect it, and "application" is an ambiguous word, and `ErrorsUnexpectedByApplication`...)

However, with so much thinking about the name, I did realize that this does not scan in an english-readability way. I would not say "with log handling failures as F" in english, I'd say "with failures handled as"; this has the additional ergonomic feature that auto-completion tools will present `log.failure` and `log.failuresHandled` next to each other, so I am going to change the name after all.

**Comment by adiroiban** [comment]
It might be due to this

> glyph requested a review from twm 2 weeks ago

I think that as long as there are requested reviewers, it is not yet considered ready for merge.

**Comment by glyph** [comment]
> I think that as long as there are requested reviewers, it is not yet considered ready for merge.

OK, I can remove twm then

**Comment by glyph** [comment]
Oh… but that doesn't help, because… somehow you reviewed *as* you, and not "on behalf of twisted-contributors"? Hmmm. I guess I should just mention people and never use the review-request mechanism.

**Comment by glyph** [comment]
siiiiiigh https://www.dockerstatus.com/pages/incident/533c6539221ae15e3f000031/6674ab18f483b5178cf6bdab

**Comment by glyph** [comment]
Not going to wait for Docker to resolve their incident here, but if there's a problem we can revert.

**Comment by adiroiban** [comment]
> Oh… but that doesn't help, because… somehow you reviewed as you, and not "on behalf of twisted-contributors"? Hmmm. I guess I should just mention people and never use the review-request mechanism.

The review robot is not AI.
It has a fixed set of rules, defined at https://github.com/chevah/github-hooks-server/blob/main/chevah/github_hooks_server/handler.py

The main use case for the robot is to allow non Twisted Org members to manage add the review label

**Comment by glyph** [comment]
> The main use case for the robot is to allow non Twisted Org members to manage add the review label

Yeah, I should contribute back to it a bit more :).  But I really want in-org members to use it as much as possible, so we have a common workflow.

**Comment by adiroiban** [comment]
> Yeah, I should contribute back to it a bit more :)

You can start by submitting bug reports and defining the rules/requirements/specifications for the robot.

----------

I think that the robot works.

For this PR, you have explicitly requested a review from Tom and Tom has not reviewed the PR yet.
So I think that the PR should not be considered ready to merge. 


---------


For example see https://github.com/twisted/twisted/pull/12213 

The PR was started as draft. As soon as the PR was no longer a draft the review label was added and a review was requested for the whole team.

I have reviewed the change and the PR was marked for merge.

Another example is https://github.com/twisted/twisted/pull/12216 ... I have  expliclty requested the review, the label was autoatically applied and the Twisted team was automatically added to review.

As soon as someone from the team has reviewd and approved the PR, it was marked for merge.


Another example is https://github.com/twisted/twisted/pull/12202 ... marked as no longer a draft, the review was automatically request, I have requested changes and the label was automatically updated.

**Comment by rnixx** [comment]
kivy (https://github.com/kivy/kivy) supports the use of twisted reactor. Therefor it uses the `ThreadedSelectReactor`. It gets installed without signal handling. When shutting down the reactor, `reactor._mainLoopShutdown()` is called before `reactor.stop()`, otherwise the process gets stuck in `threading._shutdown` (https://github.com/kivy/kivy/blob/master/kivy/support.py#L230).

This changeset removes `_mainLoopShutdown()`. While the use of this private function in kivy was probably never a clean solution it's hard to find out how to properly shutdown the reactor now.

@glyph can you point me in the right direction please?

**Comment by adiroiban** [comment]
Will `_uninstallHandler()` work for you?

**Comment by rnixx** [comment]
@adiroiban Thank you. That fixes the problem.

Just one question. Why is this not done in `ThreadedSelectReactor.stop()`?

**Comment by adiroiban** [comment]
I think that the thread is stopped via the event system ... the reason... reduce the reliance on inheritance ... I don't know. I have not designed this component.

https://github.com/twisted/twisted/blob/8f0643c8f2a1e743f24b8cdbcf4aa787b56820bb/src/twisted/internet/base.py#L1154-L1156

**Comment by rnixx** [comment]
As pointed out in my initial comment, it's possible to install the reactor without signal handling, which seems to be the cause of the issue. It might make sense to handle this case in `ThreadedSelectReactor`?

**Comment by adiroiban** [comment]
>  it's possible to install the reactor without signal handling,

I think that "signal handling" means Unix signals.

The shutdown used Twisted events notification.

From what I can see, the thread pool is lazy initialized by Twisted. 
That is, if your code never uses threads, you will not have the threading reactor part active.

It looks like you want to run Twisted in the non-main thread.

I am using this code to support some end to end tests for my Twisted application ... maybe it helps

 https://github.com/chevah/compat/blob/master/src/chevah_compat/testing/threaded_reactor.py

**Comment by glyph** [comment]
> @glyph can you point me in the right direction please?

I think that the next thing to do here is to file a bug, describing the problem that you're facing in specific details. It sounds like maybe you've got more of a feature request than a bug report, but I'm not sure why you're installing without signal handlers in the first place, so perhaps it's just a bug in the way that works.

(In the future, I plan to make a more concerted effort to remove or scramble private APIs like this on a more regular basis so that downstream users take our compatibility contract more seriously, so if you notice issues like this, reach out and try to get the features you need supported :))

**Comment by glyph** [comment]
@rnixx To be clear I think your next step here is to get us to fix the bug, which is that the reactor should not break like this when you start and stop it normally. We may need to push some work off to the application, but hopefully we can just figure out a way to do it without signals. If we *do* need to push work onto the application it should be part of the contract of starting without signals, something obvious that comes from that API like an object that gets returned to you with the relevant hook, rather than providing a public version of an API which can arbitrarily mangle internal process state to work around bugs that we have shipped.

**Pull request merged**
